http/typescript/fetch/docs/design/isomorphic.md

```diff
@@ -0,0 +1,148 @@
+        "./dist/es/src/middlewares/middlewareFactory.js": "./dist/es/src/middlewares/browser/middlewareFactory.js",
+        "./dist/cjs/src/middlewares/middlewareFactory.js": "./dist/es/src/middlewares/browser/middlewareFactory.js"
+}
+```
+
+The applications or library using `kiota-http-fetch library` will have bundlers that use this `package.json browser` mapping if the target environment is set to browser.
+
+
+## Use of `lib:dom` and `@types/node`
+
+Typescript isomorphic libraries often rely on the DOM definitions and `@types/node` as compile-time dependencies. 
+
+Example 1:
+```json
+// tsconfig.json in the isomorphic library
+
+{
+    "compileroptions": {
+        "lib":"dom"
+    }
+}
+```
+Issues with example 1: 
+    - This configuration makes DOM definitions globally available as a compile time dependency during development and it is not shipped with the library. However, when a node only application uses the isomorphic library, the DOM definitions will not be available and this causes Typescript  compile time errors.
+    - [Issue example: Removed dom lib dependency](https://github.com/prisma-labs/graphql-request/issues/26)
+    - [PR example: azure core-http](https://github.com/Azure/azure-sdk-for-js/pull/7500)
+    
+
+Example 2:
+```
+// example of use of @type/nodes
+const readableStream : NodeJS.ReadableStream
+
+or
+
+global ambient declarations at the top of the file
+/// <reference lib="dom" /> 
+/// <reference lib="node" /> 
+```
+
+Issues with example 2:
+ - When transpiling a direct reference to `@types/node`, such as  `NodeJS.ReadableStream`, the transpiled code contains a global ambient declaration `/// <reference lib="node" />`. The global ambient declarations leak the typings to the user code. 
+ - Example of issues:
+    - [Removes reference to node typings and stubs dom interfaces](https://github.com/aws/aws-sdk-js/pull/1228) 
+    - [Typings polutes global space with DOM types ](https://github.com/node-fetch/node-fetch/issues/1285)
+
+
+Other issues:
+Typescript is yet capture some of complete dynamic support that JavaScript provides. An example of this is [Node browser resolution strategy](https://github.com/microsoft/TypeScript/issues/7753) since today the compiler cannot type check based on the environment.
+
+### Solution to this:
+
+Approaches for this problem as observed in other isomorphic libraries:
+- A good reference to a very optimal solution is here: [PR example: azure core-http](https://github.com/Azure/azure-sdk-for-js/pull/7500)
+- A common approch that is taken is to create empty interfaces or shims. Then, if the user includes dom or the required definitions, the interfaces are merged. Reference [Removes reference to node typings and stubs dom interfaces](https://github.com/aws/aws-sdk-js/pull/1228) 
+
+##### DOM shims and Fetch definitions
+The `kiota-http-fetch library` relies on the DOM definitions such as `RequestInit`, `RequestInfo`. 
```

Comment:
- The use of isomorphic fetch library and defining the dom shims is independent. The isomorphic-fetch libraries do not provide these definitions. 
- For example, cross-fetch and node-fetch ambiently refer to the lib `dom` and we get the definitions from this.Here is an example of an [issue reported with this approach](https://github.com/lquixada/cross-fetch/issues/70). Defining the fetch definitions and shims is an approach taken to avoid this commonly reported issue. I referred[ to this approach azure sdks in JS have taken to solve this problem]. Also, azure replaced cross-fetch with node-fetch in their core libraries. (https://github.com/microsoft/kiota/blob/af9a3c8e3e9cd84f42a47a1f6db32e6fb94266ac/http/typescript/fetch/docs/design/isomorphic.md#use-of-libdom-and-typesnode)

We should be careful about our dependent libraries leaking the definitions. One way to avoid is to make node-fetch an optional peer dependency in the future. 

- The cross-fetch library exports the node-fetch/ browser fetch library depending on the environment which is what we are doing. There is not much of an advantage in using the isomorphic library. It gives us more control in handling the isomorphic fetch on the go. 
- The cross-fetch library does not work in service workers. This bug gets imported by the user library and I want to avoid that.

- Dom shims will need to be provided for non-fetch dom definitions too.

Reply:
I get the fact that cross fetch has a number of issues, hence why I was listing out alternatives in my original comment.
My point being: by not delegating the "runtime isomorphism" aspect to a dependency, we end up either bringing a lot of complexity into our packages or pushing some of the complexity to consumers or both.
For the "compile time isomorphism", hopefully the library we'd choose would solve for that, if it doesn't, as you mentioned we can go with type unions, shims, a combination of those etc...

Reply:
One more reason I forgot to add on why to specify these definitions is that a consumer using a custom-fetch would not want to use the isomorphic library for the definitions. 
 
The custom definitions are a super set for the DOM definitions so the goal is that a consumer can simply pass their input and no complexity pushed on them.